File: 25.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        nh = ListNode(0) # new head
        nh.next = head
        buf = [head] * k
        q, p = nh, nh.next
        while p is not None:
            # p - first node of current k-group
            # q - last node before current k-group
            for i in range(k):
                if p is None: break
                buf[i] = p
                p = p.next
            else:
                for j in range(k):
                    q.next = buf[k - 1 - j]
                    q = q.next
                q.next = p
            # less than k nodes left
            # break
        return nh.next

File: 25.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode nh = new ListNode(0); // new head
        nh.next = head;
        ListNode[] buf = new ListNode[k];
        ListNode p, q, s, t;
        int i, j;
        q = nh;
        p = q.next;
        while (p != null) {
            // p - first node of current k-group
            // q - last node before current k-group
            for (i = 0; p != null && i < k; i++) {
                buf[i] = p;
                p = p.next;
            }
            if (i < k) // left less than k nodes
                break;
            for (j = 0; j < k; j++) { // swap
                q.next = buf[k - j - 1];
                q = q.next;
            }
            q.next = p;
        }
        return nh.next;
    }
}